Support Blood Cubic Zirconia

dsimich opened the pull request:
# Description

Catching up on IOTMs that don't have support yet

Blood Cubic Zirconia:
- Will only use substats up to how much we can
- Casts all autoscend useful skills

## How Has This Been Tested?

Mobius ring chooses paradoxes as expected. Fights time cops. Summons a Wave in the Themthar Hills

## Checklist:

- [x] My code follows the style guidelines of this project.
- [x] I have performed a self-review of my own code.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have based my pull request against the [main branch](https://github.com/loathers/autoscend/tree/main) or have a good reason not to.
- [ ] I have updated the GitHub wiki [path](https://github.com/loathers/autoscend/wiki/Path-Support) or [IOTM](https://github.com/loathers/autoscend/wiki/IOTM-Support) support pages, as appropriate.

dsimich committed:
First pass at Mobius ring

dsimich committed:
Merge branch 'main' into summer25iotms

dsimich committed:
Logging

dsimich committed:
Logging and Equip Boost

dsimich committed:
Use a matcher and it works

dsimich committed:
Remove interrupt, initial wave the zone support

dsimich committed:
Actual wave the zone support

dsimich committed:
fix numeric modifier function

dsimich committed:
Use banish/instakill. Add choice for Summon a Wave

dsimich committed:
Only Summon a Wave if we actually want it in zone

dsimich committed:
BCZ start

dsimich committed:
Initial logic for BCZ. Does not take into account level!

dsimich committed:
Typos

dsimich committed:
Add daily limits and stat limits for BCZ skills. Talk to some fish

dsimich committed:
Use BCZ skills and Talk to Some Fish

dsimich committed:
Some fixes

dsimich committed:
Initial attempt to boost pheromone cocktail so it is consumed

dsimich committed:
Shrunken Head

dsimich committed:
Use hasShrunkenHead property a la autumn-aton. SoCP support

Alium58 committed:
fix gruel name due to mafia change

Alium58 committed:
fix number of free candy fights left

Caused failure to nightcap. Thought we still had free advs to spend. Don't understand original condition... am I breaking something by moving it??

Alium58 commented:
Fixed issue with failure to nightcap in latest commit. Script thought we still had free advs to spend. Don't understand original condition... am I breaking something by moving it??

dsimich committed:
Minor updates

dsimich committed:
Merge branch 'summer25iotms' of https://github.com/loathers/autoscend into summer25iotms

Alium58 committed:
Merge branch 'main' into summer25iotms

Alium58 committed:
forgot a spot to fix gruel name

midgleyc committed:
fix: check if we want to BCZ Geyser before using it

midgleyc committed:
fix: need club equipped in weapon slot for seals

midgleyc committed:
feat: also cap refract and sweat bullets at 6 per day

They could both go up more if we're above level 13, but I think we'll
lose more than we could save if we level ourselves down.

midgleyc committed:
Merge remote-tracking branch 'origin/main' into summer25iotms

midgleyc committed:
fix: mobius corrections in merge

midgleyc committed:
fix: no longer getting skele consumables either

midgleyc committed:
fix: put line endings in auto_choice_adv back to how they were

midgleyc committed:
Merge remote-tracking branch 'origin/main' into summer25iotms

midgleyc committed:
fix: check seadentwave boolean

midgleyc committed:
Merge remote-tracking branch 'origin/main' into summer25iotms

midgleyc committed:
fix: don't shrunken head again

crowkol commented:
I've run this a few ascensions recently.  Last run it made it to the hedgemaze day two hardcore.  No issues and faster than master.

kol-Kalebb commented:
Just finished running this branch again as both AT and DB in HC standard. Works much better with the level based cast restrictions, especially for sweat bullets.

kol-Kalebb committed:
BCZ support (#1714)

* Don't throw darts at the naughty sorceress (#1710)

* Don't throw darts at the naughty sorceress

* Explanatory comment

* feat: use mafia functions for mobius strip NC prediction (#1712)

The previous implementation would have always thought the NC was
available.

* Add familiar equipment override setting (#1711)

* Update any.dat

* Update autoscend_settings.txt

* Update assemble.sh

backwards compatibility with bash 3.2

* Update auto_equipment.ash

* Update any.dat

* Update autoscend_settings.txt

---------

Co-authored-by: Chris Midgley <chris.midgley@dunecomputers.co.uk>

* Use 3 heartstone skills (#1713)

This uses heartstone LUCK (as first choice for lucky use), BUFF (when asking for buffed stats, noticeably for the tower) and PALS in the familiar exp provider and the resistance provider.

* enable bcz in codpiece and enable feesh refact zones

* fix syntax error

* missed opperator

* revert monodent use with feesh. not respecting refract use for some reason

* attempt to re-enable monodent bcz synergy

* foo

* attempt to have better bcz cast scaling and fix refracted gaze

* add missed method to header

* missed save

* add guano Junction as a place to refract

* more refinements on use of feeshfract

* rework bcz casts for offstats below 70

* better limits on low off stats

* better tuning when we want to feesh/refract

* make sure we can actually spare the stats when we attempt to cast refracted gaze

---------

Co-authored-by: Alistair Crook <acrook@acrook.co.uk>
Co-authored-by: Chris Midgley <chris.midgley@dunecomputers.co.uk>
Co-authored-by: Haruto6561 <117830198+Haruto6561@users.noreply.github.com>
Co-authored-by: HippoKing <83581028+HippoKingKoL@users.noreply.github.com>

midgleyc committed:
Merge branch 'main' into summer25iotms

kol-Kalebb committed:
Bcz cleanup (#1717)

*  Added method to consistenly select correct equipment when we want to cast a bcz skill.

Cleaned up freekill equip logic and moved it to it's own method.

* fix syntax errors

* fix missing paren

* revert accidental change with sources counts

* clean up some comments

---------

Co-authored-by: Chris Midgley <midgleyc@gmail.com>
Co-authored-by: Chris Midgley <chris.midgley@dunecomputers.co.uk>

---------

Co-authored-by: Alistair Crook <acrook@acrook.co.uk>
Co-authored-by: Chris Midgley <chris.midgley@dunecomputers.co.uk>
Co-authored-by: Haruto6561 <117830198+Haruto6561@users.noreply.github.com>
Co-authored-by: HippoKing <83581028+HippoKingKoL@users.noreply.github.com>
Co-authored-by: SariusSkelrets <68650913+SariusSkelrets@users.noreply.github.com>
Co-authored-by: Danre98 <103865479+Danre98@users.noreply.github.com>
Co-authored-by: Chris Midgley <midgleyc@gmail.com>

midgleyc committed:
Merge branch 'main' into summer25iotms

midgleyc committed:
docs: add comment for why club back is last

midgleyc committed:
Merge branch 'main' into summer25iotms

midgleyc committed:
fix: add power pill back

midgleyc committed:
fix: function name change

midgleyc committed:
fix: need to check BCZ is valid

midgleyc committed:
fix: don't make items we can't consume

midgleyc commented on RELEASE/scripts/autoscend/combat/auto_combat_default_stage3.ash:
```diff
@@ -718,6 +718,11 @@ string auto_combatDefaultStage3(int round, monster enemy, string text)
 	}
 
 	// Multi-round stuns
+	if(auto_wantToBCZ($skill[BCZ: Blood Geyser]) && canUse($skill[BCZ: Blood Geyser]) && enemy_la <= 150 && !canSurvive(30.0))
```
```suggestion
	if(auto_wantToBCZ($skill[BCZ: Blood Geyser]) && canUse($skill[BCZ: Blood Geyser]) && enemy_la <= 150 && !canSurvive(5.0))
```
A lot of the time, we can't survive 30 turns, but we can kill the enemy in 3 turns so we wouldn't need to stun it.

So reduce the canSurvive requirement.

Possibly also move this down because it reduces stat points, so we should prefer other stuns.

midgleyc commented on RELEASE/scripts/autoscend/combat/auto_combat_util.ash:
```diff
@@ -739,6 +739,12 @@ string banisherCombatString(monster enemy, location loc, boolean inCombat)
 	}
 
+	if(inCombat ? (auto_have_skill($skill[Mark Your Territory]) && (!(used contains "Mark Your Territory")))
+	: auto_is_valid($skill[Mark Your Territory]) && (auto_have_skill($skill[Mark Your Territory]) || (available_amount($item[pheromone cocktail]) > 0 && inebriety_left() > 1 && !isActuallyEd()) ))
```
```suggestion
	: auto_is_valid($skill[Mark Your Territory]) && (auto_have_skill($skill[Mark Your Territory]) || (available_amount($item[pheromone cocktail]) > 0 && canDrink($item[pheromone cocktail]) && inebriety_left() > 1 && !isActuallyEd()) ))
```

midgleyc committed:
fix: ensure we can drink a pheromone cocktail before deciding to

midgleyc committed:
feat: reduce can survive requirement for blood geyser from 30 to 5

midgleyc committed:
fix: prefer non-stat-taking stuns to geyser

midgleyc committed:
feat: in amw don't cast more than 5 times

This is 411 meat which seems a worthy tradeoff.

midgleyc committed:
Merge branch 'main' into summer25iotms
